chemprop/data/utils.py

```diff
@@ -386,6 +388,22 @@ def get_data(path: str,
             if len(all_smiles) >= max_data_size:
                 break
 
+        # Precompute feature generators in batch
+        if precompute_features and features_generator is not None:
+            features_data = []  # needed since features_data is checked if none in MoleculeDatapoint construction below
+            for fg in features_generator:
+                generator = get_features_generator(fg)
+                gen_data = generator(all_smiles).reshape((len(all_smiles), -1))
+                
+                for i, _ in enumerate(all_smiles):
```

Comment:
```suggestion
                for i in range(len(all_smiles)):
```
avoid using unnamed variables when possible!

Reply:
Thanks for the feedback and for helping me along in this PR!

>I'm also not clear on the tests/test_utils.py file? Why have these functions been placed in a separate file, unless I'm missing something and they're used in more than one file. 

I moved my integration tests from tests/test_integration.py into tests/test_features.py as you recommended above. The functions in test_utils.py are now called in both the new tests/test_feaures.py and the existing tests/test_integration.py. I might advocate for test_integration.py to be further split into separate files as part of the migration effort to pytest?

Rereading our comments, it looks like you might have meant to advise that I move the unit tests to the new test/test_features.py file? I think I got a little confused as I had already written an integration test for this PR when you mentioned that I should move the testing over to pytest and to create a new file for the tests. Regardless, I think it's a good idea to move logical groupings of integration tests out to other files and did so here because: 1) the tests/test_integration.py file is getting rather large and 2) I didn't want the existing integration tests written with unittest to be in the same file as my new integration test written with pytest. Let me know what you think, I'm happy to change it back! (Unit tests are implemented in the tests/test_unit/test_features_generator.py file.)

I might also rename test_utils.py to better reflect that it's a file containing utility functions for other tests and not a file containing tests for the utility functions (as it's currently consistently named with the other test files like test_features.py, test_integration.py, etc.).

>Also not clear why these functions were named with a preceding _ if they're meant to be globally visible/used?

The functions named with a preceding _ are meant to be private. _create_raw_predict_args, _create_raw_train_args, etc. are only used within that file by the other, otherwise publicly accessible functions (train(), interpret(), etc.).